### PR TITLE
Add check for expired user-capability junctions

### PR DESCRIPTION
**Pull request by KernelDeimos**
This PR adds a predicate to CapabilityAuthService to filter expired CapabilityUserJunctions, and also adds corresponding test cases.

### CODE REVIEW

File: src/tests [hunk 1]
```diff
@@ -45,7 +45,7 @@ p({"class":"foam.nanos.test.Test","passed":1,"failed":0,"id":"ValidateTest","las
...
 // Uncomment the line below when CRUNCH is mature enough to replace our current system.
-// p({"class":"foam.nanos.crunch.crunchtest.CapabilityTest","id":"CapabilityTest","description":"test capability and junctions related"})
+p({"class":"foam.nanos.crunch.crunchtest.CapabilityTest","id":"CapabilityTest","description":"test capability and junctions related"})
```
Comment (TW80000) [inline src/tests:47]: Please remove the comment above.

File: src/foam/nanos/crunch/crunchtest/CapabilityTest.java [hunk 2]
```diff
@@ -201,6 +202,50 @@ public void testCapabilityAuthService(X x) {
...
+
+    User user = new User();
+    user.setFirstName("TestCapabilityExpireUser");
+    user.setId(9001);
+    user = (User) userDAO.put_(x, user);
+
+    AuthService auth = (AuthService) x.get("auth");
+
+    String permission1 = "permission.crunch.*";
+    cas = new CapabilityAuthService();
```
Comment (TW80000) [inline src/foam/nanos/crunch/crunchtest/CapabilityTest.java]: Unused

**Comment by kgrgreer** [comment]
How do users renew expired capabilities?

File: src/foam/nanos/crunch/crunchtest/CapabilityTest.java [hunk 3]
```diff
@@ -201,6 +202,50 @@ public void testCapabilityAuthService(X x) {
...
   }
 
+  public void testCapabilityExpires(X x) {
+
+    User user = new User();
+    user.setFirstName("TestCapabilityExpireUser");
+    user.setId(9001);
+    user = (User) userDAO.put_(x, user);
+
+    AuthService auth = (AuthService) x.get("auth");
```
Comment (TW80000) [inline src/foam/nanos/crunch/crunchtest/CapabilityTest.java:212]: How do we know that this is actually decorated with CapabilityAuthService? It's commented out in the service script for the "auth" service but I can't find anywhere that it's added or mocked out for this test.

File: src/foam/nanos/auth/CapabilityAuthService.js [hunk 4]
```diff
@@ -59,16 +61,32 @@ foam.CLASS({
...
-            EQ(UserCapabilityJunction.SOURCE_ID, user.getId()),
+            capabilityScope,
             EQ(UserCapabilityJunction.TARGET_ID, permission),
             EQ(UserCapabilityJunction.STATUS, CapabilityJunctionStatus.GRANTED)
           )) != null ) return true;
         
-        List<UserCapabilityJunction> userCapabilityJunctions = ((ArraySink) user.getCapabilities(x).getJunctionDAO().where(EQ(UserCapabilityJunction.SOURCE_ID, user.getId())).select(new ArraySink())).getArray();
+        List<UserCapabilityJunction> userCapabilityJunctions = ((ArraySink) user.getCapabilities(x).getJunctionDAO()
+          .where(
+            capabilityScope
```
Comment (TW80000) [inline src/foam/nanos/auth/CapabilityAuthService.js]: This doesn't need to be on its own line.

File: src/tests (hunk 1 above)
Comment (KernelDeimos) [reply src/tests:47]: Actually I committed this change by accident. Should we enable this now or later?

File: src/foam/nanos/crunch/crunchtest/CapabilityTest.java (hunk 3 above)
Comment (KernelDeimos) [reply src/foam/nanos/crunch/crunchtest/CapabilityTest.java:212]: None of the tests work until the decorator is enabled in services.jrl

File: src/foam/nanos/crunch/crunchtest/CapabilityTest.java (hunk 2 above)
Comment (KernelDeimos) [reply src/foam/nanos/crunch/crunchtest/CapabilityTest.java]: Updated

File: src/foam/nanos/auth/CapabilityAuthService.js (hunk 4 above)
Comment (KernelDeimos) [reply src/foam/nanos/auth/CapabilityAuthService.js]: Updated

File: src/foam/nanos/auth/CapabilityAuthService.js [hunk 5]
```diff
@@ -59,16 +61,30 @@ foam.CLASS({
         Capability cap = (Capability) capabilityDAO.find(permission);
         if ( cap != null && cap.isDeprecated(x) ) return getDelegate().checkUser(x, user, permission);
 
+        Date now = new Date();
```
Comment (kgrgreer) [inline src/foam/nanos/auth/CapabilityAuthService.js:64]: Why do we need this? Won't the cron job be expiring capabilities?

Comment (KernelDeimos) [reply src/foam/nanos/auth/CapabilityAuthService.js:64]: Yes, I think this PR conflicts with Ruby's changes. I'll ask her to confirm and then we can close this.

Comment (KernelDeimos) [reply src/foam/nanos/auth/CapabilityAuthService.js:64]: Update: This branch is now compatible with Ruby's addition of CapabilityJunctionStatus.EXPIRED

**Pull request merged**